sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -706,7 +707,22 @@ pi_result _pi_queue::resetCommandList(pi_command_list_ptr_t CommandList,
   ZE_CALL(zeFenceReset, (CommandList->second.ZeFence));
   ZE_CALL(zeCommandListReset, (CommandList->first));
   CommandList->second.InUse = false;
-  CommandList->second.EventList.clear();
+
+  // TODO: remove this unlock!!!
+  PiQueueMutex.unlock();
+
+  // Finally release/cleanup all the events in this command list.
+  // NOTE: only those that were not explicitly waited before need
+  // this handling.
+  auto &EventList = CommandList->second.EventList;
+  PI_CALL(piEventsWait(EventList.size(), EventList.data()));
```

Comment:
We unlocked PiQueueMutex above, so EventList of this Queue can be changed by another thread.
Can I suggest to use recursive mutex for now? https://github.com/againull/llvm/commit/0284c661fc3064161f78d77871e8c4e2d27ff2cc

I will check if we can avoid recursive_mutex and upload a separate patch.

Reply:
To avoid the possible cost of "recursive_mutex" (we'd need to study this separately), I've added tracking of the already locked queue in the involved interfaces. Please take a look.